### PR TITLE
python3: Fix compilation without deprecated APIs

### PR DESCRIPTION
**Pull request by neheb**
This is based on upstream work here:

https://github.com/python/cpython/pull/3934
and
https://github.com/python/cpython/pull/39430

First patch was kept as is (just rebased) while the second was combined
into one.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

Maintainer: @commodo @jefferyto 
Compile tested: ath79 mvebu ramips, maybe others...

### CODE REVIEW

File: lang/python/python3/patches/021-openssl-deprecated.patch [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ #include "openssl/x509.h"
+ #include "openssl/x509v3.h"
+@@ -128,13 +129,13 @@ static void _PySSLFixErrno(void) {
+ #include "_ssl_data.h"
+ 
+ #if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && !defined(LIBRESSL_VERSION_NUMBER)
+-#  define OPENSSL_VERSION_1_1 1
+-#  define PY_OPENSSL_1_1_API 1
++# define OPENSSL_VERSION_1_1 1
++# define PY_OPENSSL_1_1_API 1
```
Comment (commodo) [inline lang/python/python3/patches/021-openssl-deprecated.patch]: this looks like patch-noise

File: lang/python/python3/patches/021-openssl-deprecated.patch [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ #endif
+ 
+ /* LibreSSL 2.7.0 provides necessary OpenSSL 1.1.0 APIs */
+ #if defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER >= 0x2070000fL
+-#  define PY_OPENSSL_1_1_API 1
++# define PY_OPENSSL_1_1_API 1
```
Comment (commodo) [inline lang/python/python3/patches/021-openssl-deprecated.patch]: this too looks like patch-noise

**Comment by commodo** [comment]
the original PR ( https://github.com/python/cpython/pull/3934 ) was closed with comment `OpenSSL 1.0.2 will reach EOL soon. It doesn't make sense to keep this PR.`

i'm a bit vague on the details here, but any idea what Python-upstream's solution is?

File: lang/python/python3/patches/021-openssl-deprecated.patch (hunk 1 above)
Comment (neheb) [reply lang/python/python3/patches/021-openssl-deprecated.patch]: I don't know if this is valid anymore, but it made a difference back in GCC5 days. With 1 space after '#' it worked. with 2 it didn't.

**Comment by neheb** [comment]
> 
> 
> the original PR ( [python/cpython#3934](https://github.com/python/cpython/pull/3934) ) was closed with comment `OpenSSL 1.0.2 will reach EOL soon. It doesn't make sense to keep this PR.`
> 
> i'm a bit vague on the details here, but any idea what Python-upstream's solution is?

So the patch converts the API to 1.1 while providing compatibility for 1.0.2. He closed it as 1.0.2 is EOL now. But it is needed as it converts the API to the modern one.

File: lang/python/python3/patches/021-openssl-deprecated.patch (hunk 1 above)
Comment (commodo) [reply lang/python/python3/patches/021-openssl-deprecated.patch]: ohhh right;
thanks for the reminder

**Comment by commodo** [comment]
> > the original PR ( [python/cpython#3934](https://github.com/python/cpython/pull/3934) ) was closed with comment `OpenSSL 1.0.2 will reach EOL soon. It doesn't make sense to keep this PR.`
> > i'm a bit vague on the details here, but any idea what Python-upstream's solution is?
> 
> So the patch converts the API to 1.1 while providing compatibility for 1.0.2. He closed it as 1.0.2 is EOL now. But it is needed as it converts the API to the modern one.

ack
thanks for the info

**Comment by commodo** [comment]
let's wait a few days for @jefferyto to see if he has any comments

File: lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
++    else if (ctx == NULL) {
+         _setSSLError(NULL, 0, __FILE__, __LINE__);
+         return NULL;
+     }
+@@ -5241,7 +5315,7 @@ PySSL_RAND(int len, int pseudo)
+     if (bytes == NULL)
+         return NULL;
+     if (pseudo) {
+-        ok = RAND_pseudo_bytes((unsigned char*)PyBytes_AS_STRING(bytes), len);
++        ok = RAND_bytes((unsigned char*)PyBytes_AS_STRING(bytes), len);
```
Comment (jefferyto) [inline lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch]: https://github.com/python/cpython/pull/3943/commits/4809f4c0ccdce23051501f175641d8b2e8d512de may need to be combined into 021-openssl-deprecated.patch (see https://github.com/python/cpython/pull/3943#discussion_r143720329)?

File: lang/python/python3/patches/021-openssl-deprecated.patch [hunk 4]
```diff
@@ -0,0 +1,88 @@
+--- a/Modules/_ssl.c
++++ b/Modules/_ssl.c
+@@ -47,6 +47,7 @@ static PySocketModule_APIObject PySocketModule;
+ 
+ /* Include OpenSSL header files */
+ #include "openssl/rsa.h"
++#include "openssl/dh.h"
```
Comment (jefferyto) [inline lang/python/python3/patches/021-openssl-deprecated.patch]: I think this has [already been fixed](https://github.com/python/cpython/blob/v3.8.1/Modules/_ssl.c#L66) (https://github.com/python/cpython/pull/9347).

**Comment by jefferyto** [comment]
> the original PR ( python/cpython#3934 ) was closed with comment OpenSSL 1.0.2 will reach EOL soon. It doesn't make sense to keep this PR.
> 
> i'm a bit vague on the details here, but any idea what Python-upstream's solution is?

My guess is they plan to drop 1.0.2 support completely.

File: lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch (hunk 3 above)
Comment (neheb) [reply lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch]: I'll replace RAND_bytes. I think upstream is interested in getting rid ofn1.0.2 support.

File: lang/python/python3/patches/021-openssl-deprecated.patch (hunk 1 above)
Comment (neheb) [reply lang/python/python3/patches/021-openssl-deprecated.patch]: Just tested. Makes no difference anymore. Removing.

File: lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch (hunk 3 above)
Comment (neheb) [reply lang/python/python3/patches/020-ssl-module-emulate-tls-methods.patch]: I decided to modify the original patch. It's closed upstream anyway.

**Comment by neheb** [comment]
Comments were addressed.

**Comment by neheb** [comment]
> > the original PR ( [python/cpython#3934](https://github.com/python/cpython/pull/3934) ) was closed with comment OpenSSL 1.0.2 will reach EOL soon. It doesn't make sense to keep this PR.
> > i'm a bit vague on the details here, but any idea what Python-upstream's solution is?
> 
> My guess is they plan to drop 1.0.2 support completely.
I think so too.

File: lang/python/python3/patches/021-openssl-deprecated.patch (hunk 4 above)
Comment (neheb) [reply lang/python/python3/patches/021-openssl-deprecated.patch]: Removed.

File: lang/python/python3/patches/021-openssl-deprecated.patch (hunk 2 above)
Comment (neheb) [reply lang/python/python3/patches/021-openssl-deprecated.patch]: Removed. No longer an issue.

**Comment by commodo** [comment]
Let's have this in.
Personally, I would wait until Python fixes this upstream, so that we don't keep patches here.

But I do remember about the effort to reduce deprecated OpenSSL APIs; so I wouldn't want to get in the way of that [since it is a broader effort].

**Pull request merged**